### PR TITLE
Github Action for NPC and DB Validation

### PR DESCRIPTION
**Pull request by Akkarinage**
* **Addressed Issue(s)**: Lack of automated testing

* **Server Mode**: Both

* **Description of Pull Request**: 

Enables Github Actions to validate NPC and database changes.
This Action will only run if changes have been made to *either* npc/ or db/

skip-ci

### CODE REVIEW

File: .github/workflows/npc_db_validation.yml [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+          DB_RAG: 'ragnarok'
+          # MySQL (v8.0.26) provided by Github Actions is pre-configured to use root as the password.
+          # We default to 'ragnarok' for database, user and password, so no need for additional variables in this script.
+        run: |
+          mysql -u $DB_ROOT -p$DB_ROOT -e "CREATE DATABASE $DB_RAG;"
+          mysql -u $DB_ROOT -p$DB_ROOT -e "CREATE USER '$DB_RAG'@'$DB_RAG' IDENTIFIED BY '$DB_RAG';"
+          mysql -u $DB_ROOT -p$DB_ROOT -e "GRANT SELECT,INSERT,UPDATE,DELETE,CREATE ON $DB_RAG.* TO '$DB_RAG'@'$DB_RAG';"
+      
+      - name: Create MySQL User and Setup Tables
+        run: chmod +x ./tools/ci/import_sql.sh && ./tools/ci/import_sql.sh
```
Comment (Lemongrass3110) [inline .github/workflows/npc_db_validation.yml:67]: Is the chmod +x really needed?
If we did not set it correctly please change the file's chmod inside the git repo.

File: .github/workflows/npc_db_validation.yml [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+          
+      - name: Command - configure
+        env:
+            CONFIGURE_FLAGS: 'CC=gcc-${{ matrix.gcc }} CXX=g++-${{ matrix.gcc }} --enable-packetver=${{ matrix.packetver }} --enable-prere=${{ env.PRERE }} --enable-buildbot=yes'
+        run: ./configure $CONFIGURE_FLAGS
+      
+      # npc.sh enables all NPC scripts in the custom and test folders.
+      - name: Enable All NPCs for Testing
+        run: chmod +x ./tools/ci/npc.sh && ./tools/ci/npc.sh
```
Comment (Lemongrass3110) [inline .github/workflows/npc_db_validation.yml:76]: Is the chmod +x really needed?
If we did not set it correctly please change the file's chmod inside the git repo.

File: .github/workflows/npc_db_validation.yml [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    # It's a catch-all incase a Pull Request has been opened and someone is on auto-pilot.
+    if: "!contains(github.event.head_commit.message, 'ci-skip')"
+    runs-on: ${{ matrix.os }}
+    strategy:
+      matrix:
+          # The ubuntu-latest label currently points to ubuntu-18.04.
+          # Available: ubuntu-20.04, ubuntu-18.04	
+          os: [ubuntu-latest]
+          # The packet version to build the server for.
+          packetver: ['20211103']
```
Comment (Lemongrass3110) [inline .github/workflows/npc_db_validation.yml:25]:
```suggestion
```
Remove so it automatically uses latest

File: .github/workflows/npc_db_validation.yml [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+          mysql -u $DB_ROOT -p$DB_ROOT -e "CREATE DATABASE $DB_RAG;"
+          mysql -u $DB_ROOT -p$DB_ROOT -e "CREATE USER '$DB_RAG'@'$DB_RAG' IDENTIFIED BY '$DB_RAG';"
+          mysql -u $DB_ROOT -p$DB_ROOT -e "GRANT SELECT,INSERT,UPDATE,DELETE,CREATE ON $DB_RAG.* TO '$DB_RAG'@'$DB_RAG';"
+      
+      - name: Create MySQL User and Setup Tables
+        run: chmod +x ./tools/ci/import_sql.sh && ./tools/ci/import_sql.sh
+          
+      - name: Command - configure
+        env:
+            CONFIGURE_FLAGS: 'CC=gcc-${{ matrix.gcc }} CXX=g++-${{ matrix.gcc }} --enable-packetver=${{ matrix.packetver }} --enable-prere=${{ env.PRERE }} --enable-buildbot=yes'
```
Comment (Lemongrass3110) [inline .github/workflows/npc_db_validation.yml:71]:
```suggestion
            CONFIGURE_FLAGS: 'CC=gcc-${{ matrix.gcc }} CXX=g++-${{ matrix.gcc }} --enable-prere=${{ env.PRERE }} --enable-buildbot=yes'
```

**Review by Lemongrass3110 — CHANGES_REQUESTED** [review]

File: .github/workflows/npc_db_validation.yml [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+
+on: [push, pull_request]
+  paths:
+    # This workflow should run when a file in either the db/ or npc/ directory has been modified.
+    - 'db/**'
+    - 'npc/**'
+
+jobs:
+  build:
+    # Github Actions checks for 'ci skip', 'skip ci', but not a hyphenated version.
```
Comment (aleos89) [inline .github/workflows/npc_db_validation.yml:15]:
```suggestion
    # Github Actions checks for '[ci skip]', '[skip ci]', '[no ci]', '[skip actions]', or '[actions skip]' but not a hyphenated version.
```
Just to sync with the other PR.

File: .github/workflows/npc_db_validation.yml [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+        run: make clean
+
+      - name: Command - make server
+        run: make server
+
+      - name: Run Once - login-server
+        run: ./login-server --run-once
+
+      - name: Run Once - char-server
+        run: ./char-server --run-once
```
Comment (aleos89) [inline .github/workflows/npc_db_validation.yml:88]:
```suggestion
      - name: Command - make map server
        run: make map
```
Just a slight speed up. I don't think it's fully necessary to build & run the character and login servers to test the npc/ and db/ directories.

File: .github/workflows/npc_db_validation.yml [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+          DB_RAG: 'ragnarok'
+          # MySQL (v8.0.26) provided by Github Actions is pre-configured to use root as the password.
+          # We default to 'ragnarok' for database, user and password, so no need for additional variables in this script.
+        run: |
+          mysql -u $DB_ROOT -p$DB_ROOT -e "CREATE DATABASE $DB_RAG;"
+          mysql -u $DB_ROOT -p$DB_ROOT -e "CREATE USER '$DB_RAG'@'$DB_RAG' IDENTIFIED BY '$DB_RAG';"
+          mysql -u $DB_ROOT -p$DB_ROOT -e "GRANT SELECT,INSERT,UPDATE,DELETE,CREATE ON $DB_RAG.* TO '$DB_RAG'@'$DB_RAG';"
+      
+      - name: Create MySQL User and Setup Tables
+        run: chmod +x ./tools/ci/import_sql.sh && ./tools/ci/import_sql.sh
```
Comment (Akkarinage) [inline .github/workflows/npc_db_validation.yml:65]:
```suggestion
        run: ./tools/ci/import_sql.sh
```

File: .github/workflows/npc_db_validation.yml [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+          
+      - name: Command - configure
+        env:
+            CONFIGURE_FLAGS: 'CC=gcc-${{ matrix.gcc }} CXX=g++-${{ matrix.gcc }} --enable-prere=${{ env.PRERE }} --enable-buildbot=yes'
+        run: ./configure $CONFIGURE_FLAGS
+      
+      # npc.sh enables all NPC scripts in the custom and test folders.
+      - name: Enable All NPCs for Testing
+        run: chmod +x ./tools/ci/npc.sh && ./tools/ci/npc.sh
```
Comment (Akkarinage) [inline .github/workflows/npc_db_validation.yml:74]:
```suggestion
        run: ./tools/ci/npc.sh
```

File: .github/workflows/npc_db_validation.yml [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+# npc_db_validation.yml
+
+# For NPC and DB validation we only need two builds: one of Renewal and one for Pre-Renewal checks.
+# NPC scripts and database files are not platform dependent, so we can achieve this validation using only a simple linux setup.
+
+on: [push, pull_request]
+  paths:
+    # This workflow should run when a file in either the db/ or npc/ directory has been modified.
+    - 'db/**'
+    - 'npc/**'
```
Comment (Akkarinage) [inline .github/workflows/npc_db_validation.yml:11]:
```suggestion
on:
  push:
    paths:
      # This workflow should run when a file in either the db/ or npc/ directory has been modified.
      - 'db/**'
      - 'npc/**'
```

File: .github/workflows/npc_db_validation.yml [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+        run: ./configure $CONFIGURE_FLAGS
+      
+      # npc.sh enables all NPC scripts in the custom and test folders.
+      - name: Enable All NPCs for Testing
+        run: ./tools/ci/npc.sh
+
+      - name: Command - make clean
+        run: make clean
+
+      - name: Command - make map server
```
Comment (aleos89) [inline .github/workflows/npc_db_validation.yml:80]:
```suggestion
      - name: Command - make map
```
Sorry, realized you wrote the command as is in the name.

Comment (Akkarinage) [reply .github/workflows/npc_db_validation.yml:80]: How did that get there...

**Review by aleos89 — APPROVED** [review]

**Pull request merged**